### PR TITLE
SecurityContext added to ActiveGate container

### PR DESCRIPTION
**Pull request by aorcholski**
All changes related to ActiveGate container:
1) the root filesystem is mounted ReadOnly.
2) security context added:
>  privileged: false
>  allowPrivilegeEscalation: false
>  readOnlyRootFilesystem: true *
>  runAsNonRoot: true
>  capabilities:
>    drop: ["all"]
* enabled by "alpha.operator.dynatrace.com/feature-activegate-readonly-fs":"true" annotation 
3) seccompProfile is set to:
>    type: RuntimeDefault 
4)  AppArmor is enabled via AG pod annotation *
>  container.apparmor.security.beta.kubernetes.io/activegate: runtime/default 
* enabled by "alpha.operator.dynatrace.com/feature-activegate-apparmor":"true" annotation

### CODE REVIEW

File: src/controllers/activegate/capability/config.go [hunk 1]
```diff
@@ -3,6 +3,18 @@ package capability
 const (
 	ActiveGateContainerName = "activegate"
 
+	ActiveGateVolumeNameGatewayConfig = "ag-lib-gateway-config"
```
Comment (mjgrzybek) [inline src/controllers/activegate/capability/config.go:6]: plz change consts names to pattern:
`ActiveGateVolumeNameGatewayConfig`-> `ActiveGateGatewayConfigVolumeName`

File: src/controllers/activegate/reconciler/statefulset/container_eec.go [hunk 2]
```diff
@@ -21,8 +21,6 @@ const extensionsRuntimeDir = "/var/lib/dynatrace/remotepluginmodule/agent/conf/r
 const activeGateInternalCommunicationPort = 9999
 
 const (
-	eecAuthToken = "auth-tokens"
```
Comment (mjgrzybek) [inline src/controllers/activegate/reconciler/statefulset/container_eec.go:24]: why is it removed?
EDIT: ok, I see, now it's `ActiveGateVolumeNameGatewayConfig`

File: src/controllers/activegate/reconciler/statefulset/statefulset.go [hunk 3]
```diff
@@ -178,6 +179,10 @@ func buildContainers(stsProperties *statefulSetProperties, extraContainerBuilder
 	return containers
 }
 
+func pointerToBool(value bool) *bool {
```
Comment (mjgrzybek) [inline src/controllers/activegate/reconciler/statefulset/statefulset.go:182]: use `address_of.Bool()` instead

File: src/kubeobjects/slice.go [hunk 4]
```diff
@@ -11,6 +11,24 @@ func MountPathIsIn(volumeMounts []corev1.VolumeMount, mountPathToCheck string) b
 	return false
 }
 
+func MountPathIsReadOnly(volumeMounts []corev1.VolumeMount, mountPathToCheck string) bool {
```
Comment (mjgrzybek) [inline src/kubeobjects/slice.go:14]: can you merge
`func MountPathIsReadOnly(volumeMounts []corev1.VolumeMount, mountPathToCheck string) bool` and `func MountPathIsReadWrite(volumeMounts []corev1.VolumeMount, mountPathToCheck string) bool` 
to single function:
`MountPathIs(volumeMounts []corev1.VolumeMount, mountPathToCheck string, readOnly bool) bool` ?

**Review by chrismuellner — DISMISSED** [review]
Can't be merged for now as it would break older Activegate versions

File: src/controllers/activegate/capability/config.go (hunk 1 above)
Comment (aorcholski) [reply src/controllers/activegate/capability/config.go:6]: we agreed to leave it as it is

**Comment by 0sewa0** [comment]
If you are brave and want to dive into the depts of ~hell~ helm, then you could make the dynakube have these annotations when `apparmor: true` in the helm values file.  😬 

We do it like this for a deployment, different annotations, same check
https://github.com/Dynatrace/dynatrace-operator/blob/master/config/helm/chart/default/templates/Common/webhook/deployment-webhook.yaml#L36-L38

File: src/controllers/activegate/reconciler/statefulset/statefulset_test.go [hunk 5]
```diff
@@ -475,3 +569,19 @@ func buildTestInstance() *dynatracev1beta1.DynaKube {
 		},
 	}
 }
+
+func buildActiveGateVolumeDirectories(readOnly bool) []string {
```
Comment (chrismuellner) [inline src/controllers/activegate/reconciler/statefulset/statefulset_test.go:573]: does this logic also exist for the implementation? It seems there we always add every volume

**Review by chrismuellner — CHANGES_REQUESTED** [review]

File: src/kubeobjects/slice.go [hunk 6]
```diff
@@ -11,6 +16,15 @@ func MountPathIsIn(volumeMounts []corev1.VolumeMount, mountPathToCheck string) b
 	return false
 }
 
+func MountPathIs(volumeMounts []corev1.VolumeMount, mountPathToCheck string, readOnly bool) bool {
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/kubeobjects/slice.go:19]: The function checks if the MountPath exists and is ReadOnly or ReadWrite.

So maybe 
>MountPathIsReadOnlyOrReadWrite(.., mountPathToCheck string, mode bool)
would be better?

ReadOnlyMountPath and ReadWriteMountPath constants are defined for 'mode' parameter to clearly state requirement.

File: src/controllers/activegate/reconciler/statefulset/statefulset_test.go (hunk 5 above)
Comment (aorcholski) [reply src/controllers/activegate/reconciler/statefulset/statefulset_test.go:573]: Right, the logic no longer exist

Comment (aorcholski) [reply src/controllers/activegate/reconciler/statefulset/statefulset_test.go]: Eventually logic is needed. Volumes created if needed (depend on StatsD and RO fs)

File: src/kubeobjects/slice.go [hunk 7]
```diff
@@ -2,6 +2,11 @@ package kubeobjects
 
 import corev1 "k8s.io/api/core/v1"
 
+const (
+	ReadOnlyMountPath  = true
+	ReadWriteMountPath = false
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/kubeobjects/slice.go:7]: ReadOnlyMountPath and ReadWriteMountPath constants are defined for 'mode' parameter of MountPathIsReadOnlyOrReadWrite function to clearly state requirement.

**Review by 0sewa0 — CHANGES_REQUESTED** [review]
You need to add an SCC so these changes work on Openshift (openshift will complain if you have SecurityContext for a pod but no SCC for it)

Example: https://github.com/Dynatrace/dynatrace-operator/blob/master/config/helm/chart/default/templates/Openshift/operator/securitycontextconstraints.yaml

**Review by 0sewa0 — DISMISSED** [review]

**Review by chrismuellner — DISMISSED** [review]

**Pull request merged**